chain/chain/src/thread_pool.rs

```diff
@@ -0,0 +1,107 @@
+use parking_lot::Mutex;
+use std::collections::VecDeque;
+use std::sync::Arc;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::time::Duration;
+use thiserror::Error;
+use thread_priority::{RealtimeThreadSchedulePolicy, ThreadBuilder, ThreadSchedulePolicy};
+
+pub(crate) trait ThreadLimit {
+    fn max_threads(&self) -> usize;
+}
+
+/// TODO: Implement `ThreadLimit` which returns the current number of tracked shards
```

Comment:
Not sure if you need this

I think you meant the "number of shard to validate" perhaps? But I'm afraid it is not easy to compute as it changes every block height.

Reply:
I would use the number of tracked shards as a limit, as number of shards to validate should always be less or equal to this. And the number of tracked shards can only change on epoch boundary.

Reply:
CV (chunk validator) nodes track 0 shards and validate between 1 and 2 shards

And CV+BP+CP (chunk validator, block producer, chunk producer) track 1 shard and validate 1-8 shards

RPC nodes, archival nodes track all shards and validate none

Reply:
validators only track a single shard (at least most of the time) but validate multiple (~7 in the benchmarking setup IIRC)

Presumably this will also need to work for RPC nodes that track all shards, in which case you may want to get the total number of shards. Just to be safe I would add 1 too ;) Are unused threads costly or can we count on the OS to just never wake them up?

Reply:
> Are you aiming to use the same pool instance for both types of tasks?

No, definitely not.